### PR TITLE
Remove "monaco-editor/react", "react-monaco-editor" from the dependency list

### PR DESCRIPTION
**Pull request by Henry8192**
# References
NOTE: PR #33 was closed due to manual error. For previous review history, please refer to the closed PR.

Junhao found unneeded dependencies for Monaco Editor. Further investigation led to believe that non-official dependencies [@monaco-editor/react](https://github.com/suren-atoyan/monaco-react) and [react-monaco-editor](https://github.com/react-monaco-editor/react-monaco-editor) can be removed if we implement our own wrapper for Microsoft's official [monaco-editor](https://github.com/microsoft/monaco-editor), which will enable us to use the latest release from Microsoft. 

# Description
## 1. Removal of [react-monaco-editor](https://github.com/react-monaco-editor/react-monaco-editor)
Current codebase has no reference of the package. 

## 2. Removal of [monaco-editor/react](https://github.com/suren-atoyan/monaco-react)
In `src/Viewer/components/Monaco/MonacoInstance.js`:
- `handleEditorWillMount`, `handleEditorDidMount`,  `beforeMount` and `onMount` do not need any parameters, so the parameters are removed.
- Instead of returning an Editor element with passed-in arguments and letting dependency to handle the configuration, we directly create a Monaco Editor instance within the code, apply the configuration and return it.

# Validation performed
Test the log viewer with the following log files on public S3 bucket:

- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/yarn-ubuntu-resourcemanager-ip-172-31-17-135.log.1.clp.zst
- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/zip-archive.zip
- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/tar-gz-archive.tar.gz
- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/LICENSE.txt.zst
- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/LICENSE.txt.gz
- https://yscope.s3.us-east-2.amazonaws.com/sample-logs/LICENSE.txt

## Steps
1. Remove `node_modules` and `package-lock.json`, and do `npm install`.
2. `npm start` and open a zst file.
3. The browser successfully shows the log. Performed scrolling and switching themes. No errors observed in the browser's inspect element page.
4. Load another IR Stream file to see if on-the-fly editor value setting is working.

### CODE REVIEW

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 1]
```diff
@@ -100,21 +95,17 @@ function MonacoInstance ({
...
+    );
 
     /**
      * Called before the monaco editor is mounted.
-     *
-     * @param {object} monaco
      */
-    function handleEditorWillMount (monaco) {
-        beforeMount(monaco);
+    const _beforeMount = () => {
```
Comment (kirkrodrigues) [inline src/Viewer/components/Monaco/MonacoInstance.js:107]: Any particular reason for the names prefixed with `_`? We used to use the `_` to indicate a private method in a class. Of course, nowadays you have truly [private methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties).

File: webpack.common.js [hunk 2]
```diff
@@ -10,9 +10,7 @@ module.exports = {
             template: path.resolve(__dirname, "src", "index.html"),
         }),
         new MiniCssExtractPlugin(),
-        new MonacoWebpackPlugin({
-            languages: ["ascii", "ini"],
-        }),
+        new MonacoWebpackPlugin({}),
```
Comment (kirkrodrigues) [inline webpack.common.js:13]: What's the reason for this change? We originally specified the required languages to shrink the build (without specifying the languages, @junhaoliao measured the build is twice as large).

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 3]
```diff
@@ -100,21 +95,17 @@ function MonacoInstance ({
...
     const editorRef = useRef(null);
-    const monacoRef = useRef(null);
+    const editorContainerRef = useRef(null);
     const timeoutRef = useRef(null);
-    const [monacoTheme, setMonacoTheme] = useState("customLogLanguageLight");
-
-    const [language, setLanguage] = useState("");
-
-    loader.config({monaco});
+    const getMonacoThemeName = (theme) => (
```
Comment (kirkrodrigues) [inline src/Viewer/components/Monaco/MonacoInstance.js:100]: Can we move this after the lifecycle methods (before/onmount/useEffect)? That's more similar to our guidelines in other languages where we put things like constructors before generic methods.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 4]
```diff
@@ -272,22 +254,23 @@ function MonacoInstance ({
...
     }, []);
 
-    const monacoOptions = {
-        "readOnly": true,
-        "renderWhitespace": "none",
-        "wordWrap": "on",
-        "scrollBeyondLastLine": false,
-    };
+    useEffect(() => {
+        if (null === editorRef.current) {
```
Comment (kirkrodrigues) [inline src/Viewer/components/Monaco/MonacoInstance.js:258]: Could we combine this with the previous `useEffect` block?

File: src/Viewer/components/Monaco/MonacoInstance.scss [hunk 5]
```diff
@@ -4,3 +4,9 @@
 // .monaco-editor-overlaymessage {
 //  display: none !important;
 // }
+
+.monaco-container {
+    display: flex;
```
Comment (kirkrodrigues) [inline src/Viewer/components/Monaco/MonacoInstance.scss:9]: The editor doesn't seem to resize with the browser anymore. @junhaoliao mentioned we likely need [`automaticLayout`](https://microsoft.github.io/monaco-editor/typedoc/interfaces/editor.IStandaloneEditorConstructionOptions.html#automaticLayout) set in the editor's options. With that set, we also don't need `display: flex` here.

File: src/App.js [hunk 6]
```diff
@@ -38,10 +38,10 @@ export function App () {
         init();
     }, []);
 
-    const switchTheme = (theme) => {
-        localStorage.setItem(LOCAL_STORAGE_KEYS.UI_THEME, theme);
-        document.getElementById("app").setAttribute("data-theme", theme);
-        setTheme(theme);
+    const switchTheme = (newTheme) => {
```
Comment (kirkrodrigues) [inline src/App.js:41]: This isn't related to this change, right? (This is fine for this PR since it's small, but typically we try to limit the blast radius of the changes in a PR so we don't change files unrelated to it, even if it's simple refactoring. It's fine to have a separate PR just for refactoring things.)

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Nice work! Some minor comments.

File: package.json [hunk 7]
```diff
@@ -28,7 +27,7 @@
     "css-unicode-loader": "^1.0.3",
     "html-webpack-plugin": "^5.5.0",
     "jszip": "^3.10.1",
-    "monaco-editor-webpack-plugin": "^7.0.1",
+    "monaco-editor-webpack-plugin": "^7.1.0",
```
Comment (junhaoliao) [inline package.json:30]: Let's explicitly add dependency "monaco-editor" here. i.e.,
```
    "monaco-editor": "^0.45.0",
```

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 8]
```diff
@@ -1,6 +1,5 @@
-import React, {useContext, useEffect, useRef, useState} from "react";
+import React, {useContext, useEffect, useRef} from "react";
 
-import Editor, {loader} from "@monaco-editor/react";
 import * as monaco from "monaco-editor";
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:3]: Instead, let's put
```
import * as monaco from "monaco-editor/esm/vs/editor/editor.api.js";
```
to avoid `webpack` packaging unneeded code.

Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js:3]: ditto: `Viewer/components/Monaco/Shortcuts.js`

**Review by junhaoliao — CHANGES_REQUESTED** [review]
More changes are identified that can help reducing the bundle size. Please help address the comments. Thanks!

File: src/Viewer/components/Monaco/MonacoInstance.js (hunk 1 above)
Comment (kirkrodrigues) [reply src/Viewer/components/Monaco/MonacoInstance.js:107]: If it's just to avoid the name conflict with the passed-in callback, how about renaming the passed in callbacks with the suffix `Callback` (e.g., `beforeMountCallback`)?

File: webpack.common.js (hunk 2 above)
Comment (junhaoliao) [reply webpack.common.js:13]: Sorry this was my oversight. @Henry8192 As you integrate above suggestion, you may also do a size comparison on your end:
```
cd path/to/project/root

# Linux: build bundle with profiling option
webpack --profile --json > stats.json
# OR Windows: build bundle with profiling option
webpack --profile --json | Out-file 'stats.json' -Encoding OEM

# now you can check size of the `dist` directory generated under project root

# you can run webpack-bundle-analyzer to show the size of each component
npx webpack-bundle-analyzer stats.json
```
Reference: https://www.npmjs.com/package/webpack-bundle-analyzer

Things to try as you optimize the bundle size:
1. Specify `features: []` in `options` passed to `new MonacoWebpackPlugin()`.
2. (This shall go into a separate PR) Try https://webpack.js.org/plugins/compression-webpack-plugin/ with algorithms `gzip` / `br`.

File: src/Viewer/components/Monaco/MonacoInstance.js (hunk 3 above)
Comment (Henry8192) [reply src/Viewer/components/Monaco/MonacoInstance.js:100]: How about moving it to the position before the `useEffect` that calls `monaco.editor.setTheme`? I referenced `gotoLine`, which is defined right before the two `useEffect`s that are actually calling `gotoLine`.

Comment (kirkrodrigues) [reply src/Viewer/components/Monaco/MonacoInstance.js:100]: * `gotoLine` (and a lot of the existing code in this repo) doesn't really follow our guidelines. If you'd like to move `gotoLine` after the lifecycle methods, in this PR, that's also fine.
* I would still say we should move it after the `useEffect` calls. Afaik, this will work fine and it stays consistent with our guidelines, reducing the cognitive load for developers who jump into the project later.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
image

On my machine, the log-level selector seems to be behind the editor map now for some reason. Is this happening for you guys, too?

Other than that, a few minor comments.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 9]
```diff
@@ -265,29 +186,79 @@ function MonacoInstance ({
...
     };
 
-    useEffect(() => {
-        window.addEventListener("keypress", handleKeyDown);
-        return () => {
-            window.removeEventListener("keypress", handleKeyDown);
-        };
-    }, []);
+    const initMonacoEditor = () => {
+        if (null === editorRef.current) {
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js:190]: Henry and I briefly discussed this yesterday and I suggested the single exit-point approach.

Would it be better to keep a single `return` statement to simply the flow? Or do we permit early return statements for failed parameter validations at the top of functions?

[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply src/Viewer/components/Monaco/MonacoInstance.js:190]: We do permit early return statements for failed parameter validations. There is also a recommended practice (I think it comes from the early Linux devs) of using returns (and in their case, gotos) to avoid a lot of nested branches.

[the remark this answers is not in the document]
Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js:190]: Right, otherwise we can maintain a status code and keep checking that status code in the middle to avoid nested-if blocks, which is particularly helpful for functions that return a status code. 

Let me put the early return rule into our coding guidelines. :)

[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply src/Viewer/components/Monaco/MonacoInstance.js:190]: Of course, it isn't a hard and fast rule and sometimes branching is more readable than early returns.

**Comment by junhaoliao** [comment]
> image
> 
> On my machine, the log-level selector seems to be behind the editor map now for some reason. Is this happening for you guys, too?
> 
> Other than that, a few minor comments.

Same observation on my end. 

https://github.com/microsoft/vscode/blob/8494a40f51580026ba241617eb6847ff585b8f70/src/vs/editor/browser/viewParts/minimap/minimap.css#L54-L56

We should change `z-index` of this container to be some number higher than 5 then. e.g. 10.
https://github.com/y-scope/yscope-log-viewer/blob/9f3081d0dea072b7fc44ba61956a0696eca19240/src/Viewer/components/StatusBar/StatusBarMenu/StatusBarMenu.scss#L1-L4

**Review by kirkrodrigues — APPROVED** [review]
For the commit message, how about this?

`Replace monaco-editor/react with our own wrapper; Remove unnecessary monaco components from the build. (#34)`

**Review by junhaoliao — APPROVED** [review]
Looks good to me.
